File: python/motosan_agent_tool/registry.py
"""Async-safe tool registry mirroring the Rust ``ToolRegistry``."""

from __future__ import annotations

import asyncio

from .tool import Tool, ToolDef


class ToolRegistry:
    """Thread-safe (via ``asyncio.Lock``) registry of named tools.

    All public methods are ``async`` to mirror the Rust API which uses
    ``tokio::sync::RwLock``.
    """

    def __init__(self) -> None:
        self._tools: dict[str, Tool] = {}
        self._lock = asyncio.Lock()

    async def register(self, tool: Tool) -> None:
        """Register a tool. Overwrites any existing tool with the same name."""
        name = tool.def_().name
        async with self._lock:
            self._tools[name] = tool

    async def get(self, name: str) -> Tool | None:
        """Get a tool by name."""
        async with self._lock:
            return self._tools.get(name)

    async def list_defs(self) -> list[ToolDef]:
        """List all tool definitions, sorted by name for determinism."""
        async with self._lock:
            defs = [t.def_() for t in self._tools.values()]
        defs.sort(key=lambda d: d.name)
        return defs

    async def len(self) -> int:
        """Number of registered tools."""
        async with self._lock:
            return len(self._tools)

    async def is_empty(self) -> bool:
        """Whether the registry is empty."""
        async with self._lock:
            return len(self._tools) == 0

    async def deregister(self, name: str) -> Tool | None:
        """Remove a tool by name. Returns the removed tool, if any."""
        async with self._lock:
            return self._tools.pop(name, None)

    async def clear(self) -> None:
        """Remove all tools from the registry."""
        async with self._lock:
            self._tools.clear()

File: typescript/src/registry.ts
import { type Tool, ToolDef } from "./tool.js";

/**
 * In-memory registry of tools, keyed by their definition name.
 * Methods are async for API compatibility with the Rust crate
 * (which may use async trait methods backed by IO).
 */
export class ToolRegistry {
  private readonly tools: Map<string, Tool> = new Map();

  async register(tool: Tool): Promise<void> {
    const name = tool.def().name;
    this.tools.set(name, tool);
  }

  async get(name: string): Promise<Tool | undefined> {
    return this.tools.get(name);
  }

  async listDefs(): Promise<ToolDef[]> {
    return Array.from(this.tools.values()).map((t) => t.def());
  }

  async size(): Promise<number> {
    return this.tools.size;
  }

  async isEmpty(): Promise<boolean> {
    return this.tools.size === 0;
  }

  async deregister(name: string): Promise<Tool | undefined> {
    const tool = this.tools.get(name);
    this.tools.delete(name);
    return tool;
  }

  async clear(): Promise<void> {
    this.tools.clear();
  }
}
